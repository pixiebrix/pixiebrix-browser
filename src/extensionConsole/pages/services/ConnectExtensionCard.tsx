/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useState } from "react";
import { useAsyncState } from "@/hooks/common";
import { getBaseURL } from "@/services/baseService";
import { useAsyncEffect } from "use-async-effect";
import { isLinked } from "@/auth/token";
import { Card } from "react-bootstrap";
import urljoin from "url-join";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faExternalLinkAlt } from "@fortawesome/free-solid-svg-icons";

const ConnectExtensionCard: React.FunctionComponent = () => {
  const [hasExtensionKey, setHasExtensionKey] = useState(true);
  const [serviceURL] = useAsyncState(getBaseURL);

  useAsyncEffect(
    async (isMounted) => {
      const hasKey = await isLinked();
      if (isMounted()) return;
      setHasExtensionKey(hasKey);
    },
    [setHasExtensionKey]
  );

  if (hasExtensionKey) {
    return null;
  }

  return (
    <Card className="mb-4">
      <Card.Header>Connect to PixieBrix</Card.Header>
      <Card.Body>
        <p>
          By linking your browser extension with PixieBrix, you&apos;ll get
          access to team features and public services.
        </p>
        {serviceURL && (
          <a
            href={urljoin(serviceURL, "extension")}
            className="btn btn-primary"
            rel="noreferrer"
            target="_blank"
          >
            <FontAwesomeIcon icon={faExternalLinkAlt} /> Open PixieBrix Website
          </a>
        )}
      </Card.Body>
    </Card>
  );
};

export default ConnectExtensionCard;
