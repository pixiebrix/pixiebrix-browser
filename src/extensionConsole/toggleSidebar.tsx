export const SIDEBAR_ID = "sidebar";
const smallScreenMediaQuery = "(max-width: 991px)";
const sidebarActiveClassName = "active"; // Used to show/hide navbar on small screen
const sidebarIconOnlyClassName = "sidebar-icon-only"; // Used to show/hide navbar on big screen

function getSidebar(): Element {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- We know it exists
  return document.querySelector(`#${SIDEBAR_ID}`)!;
}

export const toggleSidebar = () => {
  if (window.matchMedia(smallScreenMediaQuery).matches) {
    getSidebar().classList.toggle(sidebarActiveClassName);
  } else {
    document.body.classList.toggle(sidebarIconOnlyClassName);
  }
};

export const closeSidebarOnSmallScreen = () => {
  if (window.matchMedia(smallScreenMediaQuery).matches) {
    getSidebar().classList.remove(sidebarActiveClassName);
  }
};
