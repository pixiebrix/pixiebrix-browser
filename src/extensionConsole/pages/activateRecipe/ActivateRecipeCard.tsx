/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "./ActivateRecipeCard.module.scss";

import React, { useState } from "react";
import { Button, Card, Col, Row } from "react-bootstrap";
import { truncate } from "lodash";
import { useTitle } from "@/hooks/title";
import useActivateRecipeWizard from "@/activation/useActivateRecipeWizard";
import BlockFormSubmissionViaEnterIfFirstChild from "@/components/BlockFormSubmissionViaEnterIfFirstChild";
import ReduxPersistenceContext, {
  type ReduxPersistenceContextType,
} from "@/store/ReduxPersistenceContext";
import { persistor } from "@/store/optionsStore";
import { useDispatch, useSelector } from "react-redux";
import { selectRecipeHasAnyExtensionsInstalled } from "@/store/extensionsSelectors";
import useRecipeIdParam from "@/extensionConsole/pages/useRecipeIdParam";
import { useCreateMilestoneMutation, useGetRecipeQuery } from "@/services/api";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCubes, faMagic } from "@fortawesome/free-solid-svg-icons";
import useActivateRecipe from "@/activation/useActivateRecipe";
import useMilestones from "@/hooks/useMilestones";
import Form, { type OnSubmit, type RenderBody } from "@/components/form/Form";
import { type WizardValues } from "@/activation/wizardTypes";
import Alert from "@/components/Alert";
import notify from "@/utils/notify";
import blueprintsSlice from "@/extensionConsole/pages/blueprints/blueprintsSlice";
import { BLUEPRINTS_PAGE_TABS } from "@/extensionConsole/pages/blueprints/BlueprintsPageSidebar";
import { push } from "connected-react-router";
import Loader from "@/components/Loader";

const ActivateRecipeCard: React.FC = () => {
  const dispatch = useDispatch();
  const recipeId = useRecipeIdParam();
  const isReactivate = useSelector(
    selectRecipeHasAnyExtensionsInstalled(recipeId)
  );
  // Page parent component is gating this content component on isFetching, so
  // recipe will always be resolved here
  const { data: recipe } = useGetRecipeQuery({ recipeId });

  const {
    data: wizardState,
    isLoading: isLoadingWizard,
    error: wizardError,
  } = useActivateRecipeWizard(recipe);

  const activateRecipe = useActivateRecipe("extensionConsole");
  const [activationError, setActivationError] = useState<unknown>();
  const [createMilestone] = useCreateMilestoneMutation();
  const { hasMilestone } = useMilestones();

  useTitle(
    `${isReactivate ? "Reactivate" : "Activate"} ${truncate(
      recipe.metadata.name,
      { length: 15 }
    )}`
  );

  if (isLoadingWizard) {
    return <Loader />;
  }

  if (wizardError) {
    throw wizardError;
  }

  const { wizardSteps, initialValues, validationSchema } = wizardState;

  const reduxPersistenceContext: ReduxPersistenceContextType = {
    async flush() {
      await persistor.flush();
    },
  };

  const renderBody: RenderBody = ({ values, isSubmitting }) => (
    <>
      <BlockFormSubmissionViaEnterIfFirstChild />
      <Card>
        <Card.Header className={styles.wizardHeader}>
          <Row>
            <Col>
              <div className={styles.wizardHeaderLayout}>
                <div className={styles.wizardMainInfo}>
                  <span className={styles.blueprintIcon}>
                    <FontAwesomeIcon icon={faCubes} size="2x" />
                  </span>
                  <span>
                    <Card.Title>{recipe.metadata.name}</Card.Title>
                    <code className={styles.packageId}>
                      {recipe.metadata.id}
                    </code>
                  </span>
                </div>
                <div className={styles.wizardDescription}>
                  {recipe.metadata.description}
                </div>
              </div>
              <div className={styles.activateButtonContainer}>
                <Button
                  className="text-nowrap"
                  type="submit"
                  disabled={isSubmitting}
                >
                  <FontAwesomeIcon icon={faMagic} />{" "}
                  {isReactivate ? "Reactivate" : "Activate"}
                </Button>
              </div>
            </Col>
          </Row>
        </Card.Header>
        <Card.Body className={styles.wizardBody}>
          {activationError && <Alert variant="danger">{activationError}</Alert>}
          {wizardSteps.map(({ Component, label, key }) => (
            <div key={key} className={styles.wizardBodyRow}>
              <div>
                <h4>{label}</h4>
              </div>
              <Component blueprint={recipe} reinstall={isReactivate} />
            </div>
          ))}
        </Card.Body>
      </Card>
    </>
  );

  const onSubmit: OnSubmit<WizardValues> = async (values, helpers) => {
    const { success, error } = await activateRecipe(values, recipe);

    if (success) {
      notify.success(`Installed ${recipe.metadata.name}`);

      if (!hasMilestone("first_time_public_blueprint_install")) {
        await createMilestone({
          key: "first_time_public_blueprint_install",
          metadata: {
            blueprintId: recipeId,
          },
        });

        dispatch(
          blueprintsSlice.actions.setActiveTab(BLUEPRINTS_PAGE_TABS.getStarted)
        );
      }

      dispatch(push("/mods"));
    } else {
      setActivationError(error);
    }
  };

  return (
    <ReduxPersistenceContext.Provider value={reduxPersistenceContext}>
      <Form
        initialValues={initialValues}
        validationSchema={validationSchema}
        onSubmit={onSubmit}
        renderBody={renderBody}
        renderSubmit={() => null}
      />
    </ReduxPersistenceContext.Provider>
  );
};

export default ActivateRecipeCard;
