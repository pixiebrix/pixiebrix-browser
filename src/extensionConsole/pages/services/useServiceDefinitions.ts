/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { useSelector } from "react-redux";
import { type RootState } from "@/store/optionsStore";
import { useParams } from "react-router";
import { useMemo } from "react";
import { sortBy } from "lodash";
import registry from "@/services/registry";
import { useAsyncState } from "@/hooks/common";
import { PIXIEBRIX_INTEGRATION_ID } from "@/services/constants";
import { selectConfiguredServices } from "@/store/services/servicesSelectors";
import {
  type Integration,
  type IntegrationConfig,
} from "@/types/integrationTypes";
import { type UUID } from "@/types/stringTypes";

interface IntegrationDefinitions {
  serviceDefinitions: Integration[];
  activeConfiguration: IntegrationConfig | null;
  activeService: Integration | null;
  isPending: boolean;
  showZapier: boolean;
}

const ZAPIER_SLUG = "zapier";

function useServiceDefinitions(): IntegrationDefinitions {
  const configuredServices = useSelector<RootState, IntegrationConfig[]>(
    selectConfiguredServices
  );
  const { id: configurationId } = useParams<{ id: UUID }>();

  const showZapier = configurationId === ZAPIER_SLUG;

  const [serviceDefinitions, isPending, error] = useAsyncState(
    async () =>
      sortBy(
        // eslint-disable-next-line unicorn/no-await-expression-member
        (await registry.all()).filter((x) => x.id !== PIXIEBRIX_INTEGRATION_ID),
        (x) => x.id
      ),
    []
  );

  const activeConfiguration = useMemo(
    () =>
      configurationId && configurationId !== ZAPIER_SLUG
        ? configuredServices.find((x) => x.id === configurationId)
        : null,
    [configuredServices, configurationId]
  );

  const activeService = useMemo(
    () =>
      activeConfiguration
        ? (serviceDefinitions ?? []).find(
            (x) => x.id === activeConfiguration.serviceId
          )
        : null,
    [serviceDefinitions, activeConfiguration]
  );

  if (error) {
    throw error;
  }

  return {
    activeConfiguration,
    serviceDefinitions,
    activeService,
    isPending,
    showZapier,
  };
}

export default useServiceDefinitions;
