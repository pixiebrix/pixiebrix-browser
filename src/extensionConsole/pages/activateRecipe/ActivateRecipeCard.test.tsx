/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render } from "@/extensionConsole/testHelpers";
import ActivateRecipeCard from "@/extensionConsole/pages/activateRecipe/ActivateRecipeCard";
import { waitForEffect } from "@/testUtils/testHelpers";
import { screen } from "@testing-library/react";
import registerDefaultWidgets from "@/components/fields/schemaFields/widgets/registerDefaultWidgets";
import { type RegistryId } from "@/types/registryTypes";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router";
import { type ModDefinition } from "@/types/modDefinitionTypes";
import { appApiMock } from "@/testUtils/appApiMock";
import { useGetRecipeQuery } from "@/services/api";
import AsyncStateGate from "@/components/AsyncStateGate";
import { validateRegistryId } from "@/types/helpers";
import { type RecipeResponse } from "@/types/contract";
import {
  modComponentDefinitionFactory,
  defaultModDefinitionFactory,
  metadataFactory,
} from "@/testUtils/factories/modDefinitionFactories";

registerDefaultWidgets();

// XXX: Fix this mock, getting a warning that an unhandled error was caught from submitForm() TypeError: Cannot
// destructure property 'success' of '(intermediate value)' as it is undefined
const installMock = jest.fn();
const testRecipeId = validateRegistryId("@test/recipe");

jest.mock("@/activation/useActivateRecipe.ts", () => ({
  __esModule: true,
  default: jest.fn().mockImplementation(() => installMock),
}));

jest.mock("@/extensionConsole/pages/useRecipeIdParam", () => ({
  __esModule: true,
  default: jest.fn(() => testRecipeId),
}));

global.chrome.commands.getAll = jest.fn();

function setupRecipe(recipe: ModDefinition) {
  const recipeResponse: RecipeResponse = {
    config: recipe,
    updated_at: recipe.updated_at,
    sharing: {
      public: false,
      organizations: [],
    },
  };

  appApiMock
    .onGet(`/api/recipes/${encodeURIComponent(testRecipeId)}/`)
    .reply(200, recipeResponse)
    // Databases, organizations, etc.
    .onGet()
    .reply(200, []);
}

beforeEach(() => {
  appApiMock.reset();
  jest.clearAllMocks();
});

// Activate Recipe Card is always rendered when the recipe has already been found
const RecipeCard: React.FC = () => {
  const recipeState = useGetRecipeQuery({
    recipeId: testRecipeId,
  });
  return (
    <MemoryRouter>
      <AsyncStateGate state={recipeState}>
        {() => <ActivateRecipeCard />}
      </AsyncStateGate>
    </MemoryRouter>
  );
};

describe("ActivateRecipeCard", () => {
  test("renders", async () => {
    setupRecipe(defaultModDefinitionFactory());
    const rendered = render(<RecipeCard />);
    await waitForEffect();
    expect(rendered.asFragment()).toMatchSnapshot();
  });

  test("renders successfully with null services property", async () => {
    setupRecipe(
      defaultModDefinitionFactory({
        extensionPoints: [modComponentDefinitionFactory({ services: null })],
      })
    );
    const rendered = render(<RecipeCard />);
    await waitForEffect();
    expect(rendered.asFragment()).toMatchSnapshot();
  });

  test("activate mod definition with missing required mod definition options", async () => {
    const modDefinition = defaultModDefinitionFactory({
      metadata: metadataFactory({
        id: "test/blueprint-with-required-options" as RegistryId,
        name: "Mod with Required Options",
      }),
      options: {
        schema: {
          $schema: "https://json-schema.org/draft/2019-09/schema#",
          properties: {
            database: {
              $ref: "https://app.pixiebrix.com/schemas/database#",
              title: "Database",
            },
          },
          required: ["database"],
          type: "object",
        },
        uiSchema: {},
      },
      extensionPoints: [
        modComponentDefinitionFactory({
          label: "Starter Brick for Mod with Required Options",
        }),
      ],
    });
    setupRecipe(modDefinition);

    const rendered = render(<RecipeCard />);
    await waitForEffect();
    expect(rendered.asFragment()).toMatchSnapshot();
    await userEvent.click(rendered.getByText("Activate"));
    expect(screen.getByText("Database is a required field")).not.toBeNull();
  });

  test("activate mod defintiion permissions", async () => {
    const modDefinition = defaultModDefinitionFactory({
      metadata: metadataFactory({
        id: "test/blueprint-with-required-options" as RegistryId,
        name: "A Mod",
      }),
      extensionPoints: [
        modComponentDefinitionFactory({
          label: "A Starter Brick for Mod",
        }),
      ],
    });
    setupRecipe(modDefinition);

    const rendered = render(<RecipeCard />);
    await waitForEffect();
    expect(rendered.asFragment()).toMatchSnapshot();
    await userEvent.click(rendered.getByText("Activate"));
    await waitForEffect();
    expect(installMock).toHaveBeenCalledWith(
      {
        extensions: { "0": true },
        optionsArgs: {},
        services: [],
      },
      modDefinition
    );
  });

  test("user reject permissions", async () => {
    installMock.mockResolvedValue({
      success: false,
      error: "You must accept browser permissions to activate",
    });

    const modDefinition = defaultModDefinitionFactory({
      metadata: metadataFactory({
        id: "test/blueprint-with-required-options" as RegistryId,
        name: "A Mod",
      }),
      extensionPoints: [
        modComponentDefinitionFactory({
          label: "A Starter Brick for Mod",
        }),
      ],
    });
    setupRecipe(modDefinition);

    const rendered = render(<RecipeCard />);
    await waitForEffect();
    await userEvent.click(rendered.getByText("Activate"));
    await waitForEffect();

    expect(
      screen.getByText("You must accept browser permissions to activate")
    ).toBeVisible();
  });
});
