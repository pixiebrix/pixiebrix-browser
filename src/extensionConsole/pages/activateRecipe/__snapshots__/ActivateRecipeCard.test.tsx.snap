// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivateRecipeCard activate recipe permissions 1`] = `
<DocumentFragment>
  <form
    class=""
    novalidate=""
  >
    <input
      disabled=""
      hidden=""
      type="submit"
    />
    <div
      class="card"
    >
      <div
        class="wizardHeader card-header"
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="wizardHeaderLayout"
            >
              <div
                class="wizardMainInfo"
              >
                <span
                  class="blueprintIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-cubes fa-w-16 fa-2x "
                    data-icon="cubes"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M488.6 250.2L392 214V105.5c0-15-9.3-28.4-23.4-33.7l-100-37.5c-8.1-3.1-17.1-3.1-25.3 0l-100 37.5c-14.1 5.3-23.4 18.7-23.4 33.7V214l-96.6 36.2C9.3 255.5 0 268.9 0 283.9V394c0 13.6 7.7 26.1 19.9 32.2l100 50c10.1 5.1 22.1 5.1 32.2 0l103.9-52 103.9 52c10.1 5.1 22.1 5.1 32.2 0l100-50c12.2-6.1 19.9-18.6 19.9-32.2V283.9c0-15-9.3-28.4-23.4-33.7zM358 214.8l-85 31.9v-68.2l85-37v73.3zM154 104.1l102-38.2 102 38.2v.6l-102 41.4-102-41.4v-.6zm84 291.1l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6zm240 112l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                <span>
                  <div
                    class="card-title h5"
                  >
                    A Mod
                  </div>
                  <code
                    class="packageId"
                  >
                    test/blueprint-with-required-options
                  </code>
                </span>
              </div>
              <div
                class="wizardDescription"
              >
                Recipe generated from factory
              </div>
            </div>
            <div
              class="activateButtonContainer"
            >
              <button
                class="text-nowrap btn btn-primary"
                type="submit"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-magic fa-w-16 "
                  data-icon="magic"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M224 96l16-32 32-16-32-16-16-32-16 32-32 16 32 16 16 32zM80 160l26.66-53.33L160 80l-53.34-26.67L80 0 53.34 53.33 0 80l53.34 26.67L80 160zm352 128l-26.66 53.33L352 368l53.34 26.67L432 448l26.66-53.33L512 368l-53.34-26.67L432 288zm70.62-193.77L417.77 9.38C411.53 3.12 403.34 0 395.15 0c-8.19 0-16.38 3.12-22.63 9.38L9.38 372.52c-12.5 12.5-12.5 32.76 0 45.25l84.85 84.85c6.25 6.25 14.44 9.37 22.62 9.37 8.19 0 16.38-3.12 22.63-9.37l363.14-363.15c12.5-12.48 12.5-32.75 0-45.24zM359.45 203.46l-50.91-50.91 86.6-86.6 50.91 50.91-86.6 86.6z"
                    fill="currentColor"
                  />
                </svg>
                 Activate
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="wizardBody card-body"
      >
        <div
          class="wizardBodyRow"
        >
          <div>
            <h4>
              Permissions & URLs
            </h4>
          </div>
          <div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <p>
                  No special permissions required
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`ActivateRecipeCard activate recipe with missing required recipe options 1`] = `
<DocumentFragment>
  <form
    class=""
    novalidate=""
  >
    <input
      disabled=""
      hidden=""
      type="submit"
    />
    <div
      class="card"
    >
      <div
        class="wizardHeader card-header"
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="wizardHeaderLayout"
            >
              <div
                class="wizardMainInfo"
              >
                <span
                  class="blueprintIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-cubes fa-w-16 fa-2x "
                    data-icon="cubes"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M488.6 250.2L392 214V105.5c0-15-9.3-28.4-23.4-33.7l-100-37.5c-8.1-3.1-17.1-3.1-25.3 0l-100 37.5c-14.1 5.3-23.4 18.7-23.4 33.7V214l-96.6 36.2C9.3 255.5 0 268.9 0 283.9V394c0 13.6 7.7 26.1 19.9 32.2l100 50c10.1 5.1 22.1 5.1 32.2 0l103.9-52 103.9 52c10.1 5.1 22.1 5.1 32.2 0l100-50c12.2-6.1 19.9-18.6 19.9-32.2V283.9c0-15-9.3-28.4-23.4-33.7zM358 214.8l-85 31.9v-68.2l85-37v73.3zM154 104.1l102-38.2 102 38.2v.6l-102 41.4-102-41.4v-.6zm84 291.1l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6zm240 112l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                <span>
                  <div
                    class="card-title h5"
                  >
                    Mod with Required Options
                  </div>
                  <code
                    class="packageId"
                  >
                    test/blueprint-with-required-options
                  </code>
                </span>
              </div>
              <div
                class="wizardDescription"
              >
                Recipe generated from factory
              </div>
            </div>
            <div
              class="activateButtonContainer"
            >
              <button
                class="text-nowrap btn btn-primary"
                type="submit"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-magic fa-w-16 "
                  data-icon="magic"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M224 96l16-32 32-16-32-16-16-32-16 32-32 16 32 16 16 32zM80 160l26.66-53.33L160 80l-53.34-26.67L80 0 53.34 53.33 0 80l53.34 26.67L80 160zm352 128l-26.66 53.33L352 368l53.34 26.67L432 448l26.66-53.33L512 368l-53.34-26.67L432 288zm70.62-193.77L417.77 9.38C411.53 3.12 403.34 0 395.15 0c-8.19 0-16.38 3.12-22.63 9.38L9.38 372.52c-12.5 12.5-12.5 32.76 0 45.25l84.85 84.85c6.25 6.25 14.44 9.37 22.62 9.37 8.19 0 16.38-3.12 22.63-9.37l363.14-363.15c12.5-12.48 12.5-32.75 0-45.24zM359.45 203.46l-50.91-50.91 86.6-86.6 50.91 50.91-86.6 86.6z"
                    fill="currentColor"
                  />
                </svg>
                 Activate
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="wizardBody card-body"
      >
        <div
          class="wizardBodyRow"
        >
          <div>
            <h4>
              Configure Mod
            </h4>
          </div>
          <div
            class="formGroup form-group row"
          >
            <label
              class="label form-label col-form-label col-xl-2 col-lg-3"
              for="optionsArgs.database"
            >
              Database
            </label>
            <div
              class="col-xl-10 col-lg-9"
            >
              <div
                class="root"
              >
                <div
                  class="field"
                >
                  <div
                    class=" css-b62m3t-container"
                  >
                    <span
                      class="css-1f43avz-a11yText-A11yText"
                      id="react-select-2-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class=" css-13cymwt-control"
                    >
                      <div
                        class=" css-1fdsijx-ValueContainer"
                      >
                        <div
                          class=" css-1jqq78o-placeholder"
                          id="react-select-2-placeholder"
                        >
                          Select...
                        </div>
                        <div
                          class=" css-qbdosj-Input"
                          data-value=""
                        >
                          <input
                            aria-autocomplete="list"
                            aria-describedby="react-select-2-placeholder"
                            aria-expanded="false"
                            aria-haspopup="true"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            class=""
                            id="react-select-2-input"
                            role="combobox"
                            spellcheck="false"
                            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                      <div
                        class=" css-1hb7zxy-IndicatorsContainer"
                      >
                        <span
                          class=" css-1u9des2-indicatorSeparator"
                        />
                        <div
                          aria-hidden="true"
                          class=" css-1xc3v61-indicatorContainer"
                        >
                          <svg
                            aria-hidden="true"
                            class="css-tj5bde-Svg"
                            focusable="false"
                            height="20"
                            viewBox="0 0 20 20"
                            width="20"
                          >
                            <path
                              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                    <input
                      name="optionsArgs.database"
                      type="hidden"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class="dropdown dropdown"
                  data-test-selected="Database"
                  data-testid="toggle-optionsArgs.database"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="true"
                    class="dropdown-toggle btn btn-link"
                    disabled=""
                    type="button"
                  >
                    <span
                      class="symbol"
                    >
                      <test-file-stub
                        classname="root"
                      />
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="wizardBodyRow"
        >
          <div>
            <h4>
              Permissions & URLs
            </h4>
          </div>
          <div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <p>
                  No special permissions required
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`ActivateRecipeCard renders 1`] = `
<DocumentFragment>
  <form
    class=""
    novalidate=""
  >
    <input
      disabled=""
      hidden=""
      type="submit"
    />
    <div
      class="card"
    >
      <div
        class="wizardHeader card-header"
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="wizardHeaderLayout"
            >
              <div
                class="wizardMainInfo"
              >
                <span
                  class="blueprintIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-cubes fa-w-16 fa-2x "
                    data-icon="cubes"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M488.6 250.2L392 214V105.5c0-15-9.3-28.4-23.4-33.7l-100-37.5c-8.1-3.1-17.1-3.1-25.3 0l-100 37.5c-14.1 5.3-23.4 18.7-23.4 33.7V214l-96.6 36.2C9.3 255.5 0 268.9 0 283.9V394c0 13.6 7.7 26.1 19.9 32.2l100 50c10.1 5.1 22.1 5.1 32.2 0l103.9-52 103.9 52c10.1 5.1 22.1 5.1 32.2 0l100-50c12.2-6.1 19.9-18.6 19.9-32.2V283.9c0-15-9.3-28.4-23.4-33.7zM358 214.8l-85 31.9v-68.2l85-37v73.3zM154 104.1l102-38.2 102 38.2v.6l-102 41.4-102-41.4v-.6zm84 291.1l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6zm240 112l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                <span>
                  <div
                    class="card-title h5"
                  >
                    Blueprint 1
                  </div>
                  <code
                    class="packageId"
                  >
                    test/blueprint-1
                  </code>
                </span>
              </div>
              <div
                class="wizardDescription"
              >
                Recipe generated from factory
              </div>
            </div>
            <div
              class="activateButtonContainer"
            >
              <button
                class="text-nowrap btn btn-primary"
                type="submit"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-magic fa-w-16 "
                  data-icon="magic"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M224 96l16-32 32-16-32-16-16-32-16 32-32 16 32 16 16 32zM80 160l26.66-53.33L160 80l-53.34-26.67L80 0 53.34 53.33 0 80l53.34 26.67L80 160zm352 128l-26.66 53.33L352 368l53.34 26.67L432 448l26.66-53.33L512 368l-53.34-26.67L432 288zm70.62-193.77L417.77 9.38C411.53 3.12 403.34 0 395.15 0c-8.19 0-16.38 3.12-22.63 9.38L9.38 372.52c-12.5 12.5-12.5 32.76 0 45.25l84.85 84.85c6.25 6.25 14.44 9.37 22.62 9.37 8.19 0 16.38-3.12 22.63-9.37l363.14-363.15c12.5-12.48 12.5-32.75 0-45.24zM359.45 203.46l-50.91-50.91 86.6-86.6 50.91 50.91-86.6 86.6z"
                    fill="currentColor"
                  />
                </svg>
                 Activate
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="wizardBody card-body"
      >
        <div
          class="wizardBodyRow"
        >
          <div>
            <h4>
              Permissions & URLs
            </h4>
          </div>
          <div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <p>
                  No special permissions required
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
